store unencrypted size in the unencrypted_size column

icewind1991 opened the pull request:
*The column is right there*

Store the unencrypted size in the `unencrypted_size` column instead of the `size` column, fixes some issues where other code (object storage) stores the encrypted size in the cache.

icewind1991 committed:
store unencrypted size in the unencrypted_size column

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 commented:
To test:

- setup instance with s3 primary storage
- enable encryption
- upload some files, verify that the listed size is correct
- verify that the files can be downloaded a few times and the size stays correct
- keep an eye out for things exploding :)

szaimen commented on lib/public/Files/Cache/ICacheEntry.php:
```diff
@@ -162,4 +162,14 @@ public function getCreationTime(): ?int;
+	 * this might be different from the result of getSize
+	 *
+	 * @return int
+	 * @since 24.0.0
```
```suggestion
	 * @since 25.0.0
```
?

CarlSchwan replied:
I suppose this need to get backported, so probably fine

CarlSchwan commented on lib/public/Files/Cache/ICacheEntry.php:
```diff
@@ -162,4 +162,14 @@ public function getCreationTime(): ?int;
+	/**
+	 * Get the unencrypted size
+	 *
+	 * this might be different from the result of getSize
```
```suggestion
	 * This might be different from the result of getSize.
```

CarlSchwan commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -478,7 +488,7 @@ public function fopen($path, $mode) {
 	 * @return int unencrypted size
 	 */
-	protected function verifyUnencryptedSize($path, $unencryptedSize) {
+	protected function verifyUnencryptedSize(string $path, int $unencryptedSize) {
```
```suggestion
	protected function verifyUnencryptedSize(string $path, int $unencryptedSize): int {
```

While at it :)

CarlSchwan commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -510,7 +520,7 @@ protected function verifyUnencryptedSize($path, $unencryptedSize) {
 	 * @return int calculated unencrypted size
 	 */
-	protected function fixUnencryptedSize($path, $size, $unencryptedSize) {
+	protected function fixUnencryptedSize(string $path, int $size, int $unencryptedSize) {
```
```suggestion
	protected function fixUnencryptedSize(string $path, int $size, int $unencryptedSize): int {
```

CarlSchwan reviewed (approved):
Code looks :+1: 

Just some minor comments that can be ignored

PVince81 commented:
as discussed in chat:

- [ ] check if/how unencrypted_size must be propagated like size, which means it will exist on folders and might affect the logic that detects "-1" sizes in some scenarios

icewind1991 committed:
make expression build return IQueryFunction instead of string

this allows passing the expressions to further expressions without them being escaped as column names

Signed-off-by: Robin Appelman <robin@icewind.nl>

icewind1991 committed:
add case statement to sql function builder

Signed-off-by: Robin Appelman <robin@icewind.nl>

PVince81 commented on lib/private/DB/QueryBuilder/ExpressionBuilder/ExpressionBuilder.php:
```diff
@@ -114,12 +114,12 @@ public function orX(...$x): ICompositeExpression {
 		$x = $this->helper->quoteColumnName($x);
 		$y = $this->helper->quoteColumnName($y);
-		return $this->expressionBuilder->comparison($x, $operator, $y);
+		return new QueryFunction($this->expressionBuilder->comparison($x, $operator, $y));
```
is this a public  API change?

would need to be mentionned in https://github.com/nextcloud/server/issues/32117

PVince81 replied:
@icewind1991 can you clarify ?

summersab replied:
I applied this patch to my (somewhat customized) Docker image that I am developing. It worked just fine a few days ago, but when I ran it just now, it threw the following error upon trying to access the root page:
```
{
  "reqId": "EIZn0TU6px16gqETseAb",
  "level": 3,
  "time": "2022-05-30T00:09:52+00:00",
  "remoteAddr": "192.168.0.202",
  "user": "--",
  "app": "PHP",
  "method": "GET",
  "url": "/",
  "message": "Declaration of OC\\DB\\QueryBuilder\\ExpressionBuilder\\ExpressionBuilder::comparison($x, string $operator, $y, $type = null): OCP\\DB\\QueryBuilder\\IQueryFunction must be compatible with OCP\\DB\\QueryBuilder\\IExpressionBuilder::comparison($x, string $operator, $y, $type = null): string at /var/www/html/lib/private/DB/QueryBuilder/ExpressionBuilder/ExpressionBuilder.php#119",
  "userAgent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/97.0.4692.71 Safari/537.36",
  "version": "24.0.1.1"
}
```

I'm fairly adept at debugging Nextcloud and know my way around the code, but this particular patch is quite involved, so I'm not sure where to start to provide more information. Feel free to ask, though - I'm happy to help!

PVince81 commented on lib/private/Files/Cache/Cache.php:
```diff
@@ -873,8 +875,16 @@ public function calculateFolderSize($path, $entry = null) {
+					], 'unencrypted_size')
+				), 'unencrypted_sum')
+				->selectAlias($query->func()->min('unencrypted_size'), 'unencrypted_min')
+				->selectAlias($query->func()->max('unencrypted_size'), 'unencrypted_max')
```
let's hope the extra aggregation doesn't have a perf penalty

otherwise we might want to disable it when encryption is off

icewind1991 replied:
don't expect it to have any since it needs to loop over all items in the folder for the other aggregations

PVince81 commented on lib/private/Files/Storage/Wrapper/Encryption.php:
```diff
@@ -808,13 +843,6 @@ private function copyBetweenStorage(Storage\IStorage $sourceStorage, $sourceInte
-	 *
-	 * @param string $path
-	 * @return string
-	 */
```
use inherit docs instead?

icewind1991 replied:
both psalm and ide's implicitly inherit docs when none is provided

PVince81 commented:
@icewind1991 some conflicts

PVince81 reviewed (approved):
:+1:

DaphneMuller commented:
@CarlSchwan  could you please review this?

DaphneMuller commented:
@szaimen could you please review?

PVince81 commented:
oh, so the backport got merged before master: https://github.com/nextcloud/server/pull/32708

PVince81 commented:
@icewind1991 can you communicate the API change somehow and put it in https://github.com/nextcloud/server/issues/32117

CarlSchwan commented:
https://github.com/nextcloud/circles/pull/1052

ChristophWurst commented:
This also breaks the Mail app
``TypeError: OCA\Mail\Db\MailboxMapper::eqOrNull(): Return value must be of type string, OC\DB\QueryBuilder\QueryFunction returned``

ChristophWurst commented on lib/public/DB/QueryBuilder/IExpressionBuilder.php:
```diff
@@ -132,14 +132,14 @@ public function comparison($x, string $operator, $y, $type = null): string;
 	 * @psalm-taint-sink sql $type
 	 */
-	public function eq($x, $y, $type = null): string;
+	public function eq($x, $y, $type = null): IQueryFunction;
```
from a naming perspective this should have been something like an `IExpression`

icewind1991 commented:
Given the unexpected breakage I guess reverting this and using the more "minimal" version from https://github.com/nextcloud/server/pull/32708 instead.

icewind1991 commented:
https://github.com/nextcloud/server/pull/32943
